src/plans/D3MCompoundPlan.sol

```diff
@@ -0,0 +1,187 @@
+    function _wmul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = (x * y) / WAD;
+    }
+    function _wdiv(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = (x * WAD) / y;
+    }
+
+    // --- Admin ---
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    function file(bytes32 what, uint256 data) external auth {
+        if (what == "barb") {
+            barb = data;
+        } else revert("D3MCompoundPlan/file-unrecognized-param");
+        emit File(what, data);
+    }
+    function file(bytes32 what, address data) external auth {
+        if (what == "tack") tack = InterestRateModelLike(data);
+        else revert("D3MCompoundPlan/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    function getTargetAssets(uint256 currentAssets) external override view returns (uint256) {
+        uint256 targetInterestRate = barb;
+        if (targetInterestRate == 0) return 0; // De-activated
+
+        uint256 borrows = cDai.totalBorrows();
+        uint256 targetTotalPoolSize = _calculateTargetSupply(targetInterestRate, borrows);
+        uint256 totalPoolSize = cDai.getCash() + borrows - cDai.totalReserves();
+
+        if (targetTotalPoolSize >= totalPoolSize) {
+            // Increase debt (or same)
+            return currentAssets + targetTotalPoolSize - totalPoolSize;
+        } else {
+            // Decrease debt
+            unchecked {
+                uint256 decrease = totalPoolSize - targetTotalPoolSize;
+                if (currentAssets >= decrease) {
+                    return currentAssets - decrease;
+                } else {
+                    return 0;
+                }
+            }
+        }
+    }
+
+    function _calculateTargetSupply(uint256 targetInterestRate, uint256 borrows) internal view returns (uint256) {
+        uint256 kink                   = tack.kink();
+        uint256 multiplierPerBlock     = tack.multiplierPerBlock();
+        uint256 baseRatePerBlock       = tack.baseRatePerBlock();
+        uint256 jumpMultiplierPerBlock = tack.jumpMultiplierPerBlock();
+
+        uint256 normalRate = _wmul(kink, multiplierPerBlock) + baseRatePerBlock;
+
+        uint256 targetUtil;
+        if (targetInterestRate > normalRate) {
+            if (jumpMultiplierPerBlock == 0) return 0; // illegal rate, max is normal rate for this case
+            targetUtil = kink + _wdiv(targetInterestRate - normalRate, jumpMultiplierPerBlock); // (1)
+        } else if (targetInterestRate > baseRatePerBlock) {
+            targetUtil = _wdiv(targetInterestRate - baseRatePerBlock, multiplierPerBlock);      // (2)
+        } else {
+            // if (target == base) => (borrows == 0) => supply does not matter
+            // if (target  < base) => illegal rate
+            return 0;
```

Comment:
Hmmmm...not sure about this behavior. I'd naively think the way to interpret `targetInterestRate <= baseRatePerBlock` would be "inject liquidity up to available debt ceiling to make the rate as close to the target as possible". Instead, we'll unwind in such a state, maximizing the interest rate instead of minimizing it. Ofc I guess one can just argue that governance or what is setting the target interest rate should just be taking this into account, but IMO it's a counterintuitive edge case. E.g. I'd expect setting the target interest rate to zero would be equivalent to "inject max allowed liquidity".

Reply:
Yeah it makes sense. As this is also relevant for AAVE let's continue the discussion here - https://github.com/makerdao/dss-direct-deposit/pull/46/files#r893886318.